File: mtui/commands/openqa_jobs.py
"""The ``openqa_jobs`` interactive command.

Lists the individual openQA jobs for the loaded update's incident build, so a
tester (or an MCP client) can see *which* scenarios passed or failed — and judge
whether a failure relates to the package under test — rather than only the
per-version PASSED/FAILED/RUNNING summary that ``openqa_overview`` prints.

By default ``obsoleted`` jobs (superseded by a later retrigger) are dropped; only
the current run matters. Use ``--all`` to keep them, ``--failed`` to show only
non-passing jobs, and ``--arch`` to filter by architecture. Jobs that have not
finished (scheduled/running — openQA reports their result as ``none``) are
pending, not failures: ``--failed`` excludes them and the listing shows their
state in yellow instead of a red ``none``.
"""

from __future__ import annotations

from collections import Counter
from logging import getLogger

from ..cli.argparse import ArgumentParser
from ..cli.colors import green, red, yellow
from ..cli.completion import complete_choices, template_completion
from ..data_sources import oqa_search as oqa
from ..support.http import resolve_verify
from ..support.misc import requires_update
from . import Command

logger = getLogger("mtui.commands.openqa_jobs")

# openQA results that count as "not a failure" for the --failed filter.
_PASSING = frozenset({"passed", "softfailed"})
_NEUTRAL = frozenset({"obsoleted", "skipped"})
# A job that has not finished yet: openQA keeps ``result`` at ``none``
# (some API paths omit it entirely) until the job is done. Pending work
# is not a failure -- it must neither pass the --failed filter nor be
# painted red.
_PENDING = frozenset({"none", ""})


def _display_result(job) -> str:
    """The label to print/count for a job.

    An unfinished job's ``result`` is the meaningless ``none``; its
    ``state`` (``scheduled``, ``running``, ...) is what a tester wants
    to see instead.
    """
    if job.result in _PENDING:
        return job.state or "pending"
    return job.result


class OpenQAJobs(Command):
    """List the individual openQA jobs for the loaded update."""

    command = "openqa_jobs"
    scope = "fanout"

    @classmethod
    def _add_arguments(cls, parser: ArgumentParser) -> None:
        """Register the command's arguments."""
        cls._add_template_arg(parser)
        parser.add_argument(
            "--all",
            action="store_true",
            help="include obsoleted (superseded) jobs",
        )
        parser.add_argument(
            "--failed",
            action="store_true",
            help="show only non-passing jobs (failed / parallel_failed / incomplete)",
        )
        parser.add_argument(
            "--arch",
            type=str,
            default=None,
            help="only jobs for this architecture",
        )
        parser.add_argument(
            "--url-openqa",
            type=str,
            default=None,
            help="Override openQA URL (default: config openqa_instance)",
        )
        parser.add_argument(
            "--url-dashboard-qam",
            type=str,
            default=None,
            help="Override QAM Dashboard base URL (default: derived from config)",
        )

    @requires_update
    def __call__(self) -> None:
        """Fetch and print the incident's openQA jobs."""
        rrid = self.metadata.rrid
        oqa.set_verify(resolve_verify(True, self.config.ssl_verify))

        url_openqa = self.args.url_openqa or self.config.openqa_instance
        url_dashboard_qam = self.args.url_dashboard_qam or (
            self.config.qem_dashboard_api.rstrip("/").removesuffix("/api")
        )

        # incident_id is an int for Maintenance, "1.2" for SLFO; fall back to the
        # review id (the gitea PR number) in the SLFO case -- mirrors openqa_overview.
        incident_id = rrid.maintenance_id
        effective_incident_id = (
            incident_id if isinstance(incident_id, int) else rrid.review_id
        )

        try:
            build, _ = oqa.get_incident_info(url_dashboard_qam, effective_incident_id)
        except oqa._HTTPError as e:  # noqa: SLF001 -- module-internal exception
            logger.error("Failed to query QEM Dashboard: %s", e)
            return

        jobs = oqa.incident_jobs(build, url_openqa, include_obsoleted=self.args.all)
        if self.args.arch:
            jobs = [j for j in jobs if j.arch == self.args.arch]
        total = len(jobs)
        pending = sum(1 for j in jobs if j.result in _PENDING)
        if self.args.failed:
            jobs = [
                j
                for j in jobs
                if j.result not in _PASSING
                and j.result not in _NEUTRAL
                and j.result not in _PENDING
            ]

        if not jobs:
            if self.args.failed and total:
                # Jobs exist, none failed (yet). Saying "no jobs" here would
                # make an in-progress build indistinguishable from an empty
                # or all-passed one, so name the pending work explicitly.
                note = f"; {pending} of {total} still pending" if pending else ""
                self.println(yellow(f"No failed openQA jobs for build {build!r}{note}"))
            else:
                self.println(yellow(f"No openQA jobs for build {build!r}"))
            return

        counts = Counter(_display_result(j) for j in jobs)
        summary = ", ".join(f"{k}={v}" for k, v in sorted(counts.items()))
        self.println(f"openQA jobs for build {build} ({len(jobs)}): {summary}")
        self.println("")
        for j in jobs:
            if j.result in _PASSING:
                colour = green
            elif j.result in _NEUTRAL or j.result in _PENDING:
                colour = yellow
            else:
                colour = red
            self.println(
                f"  {colour(_display_result(j).ljust(15))} {j.arch:<8} {j.test}  {j.url}"
            )

    @staticmethod
    def complete(state, text, line, begidx, endidx) -> list[str]:
        """Provides tab completion for the command."""
        return complete_choices(
            [
                ("--all",),
                ("--failed",),
                ("--arch",),
                ("--url-openqa",),
                ("--url-dashboard-qam",),
                *template_completion(state),
            ],
            line,
            text,
        )

File: crates/mtui-core/src/commands/openqa_jobs.rs
//! The `openqa_jobs` command — list the individual openQA jobs for an update.

use std::collections::BTreeMap;

use async_trait::async_trait;
use clap::{Arg, ArgAction, ArgMatches};
use mtui_datasources::oqa_search as oqa;

use crate::command::{Command, Scope};
use crate::commands::support::{require_update, template_completion};
use crate::error::{CommandError, CommandResult};
use crate::session::Session;

/// openQA results that count as "not a failure" for the `--failed` filter
/// (upstream `_PASSING`).
const PASSING: &[&str] = &["passed", "softfailed"];
/// openQA results that are neutral (neither pass nor fail; upstream `_NEUTRAL`).
const NEUTRAL: &[&str] = &["obsoleted", "skipped"];
/// openQA job states that mean the job has finished; any other state
/// (`scheduled`, `assigned`, `setup`, `running`, `uploading`, ...) is pending
/// and its `result` is not yet meaningful.
const TERMINAL_STATES: &[&str] = &["done", "cancelled"];

/// Whether an openQA job has not finished yet. Pending jobs must not be counted
/// as failures by `--failed`, and are surfaced separately as "N still pending".
fn is_pending(job: &oqa::JobResult) -> bool {
    !TERMINAL_STATES.contains(&job.state.as_str())
}

/// Lists the individual openQA jobs for the loaded update's incident build.
///
/// Ports upstream `mtui.commands.openqa_jobs.OpenQAJobs`. By default `obsoleted`
/// jobs are dropped; `--all` keeps them, `--failed` shows only genuinely failed
/// jobs (pending/unfinished jobs are excluded and reported as "N still
/// pending"), and `--arch` filters by architecture. Requires a loaded update.
pub struct OpenQAJobs;

#[async_trait]
impl Command for OpenQAJobs {
    fn name(&self) -> &'static str {
        "openqa_jobs"
    }

    fn about(&self) -> Option<&'static str> {
        Some("Lists the individual openQA jobs for the loaded update's incident build.")
    }

    fn scope(&self) -> Scope {
        Scope::Fanout
    }

    fn configure(&self, cmd: clap::Command) -> clap::Command {
        cmd.arg(
            Arg::new("all")
                .long("all")
                .action(ArgAction::SetTrue)
                .help("keep obsoleted jobs (superseded by a retrigger)"),
        )
        .arg(
            Arg::new("failed")
                .long("failed")
                .action(ArgAction::SetTrue)
                .help("show only genuinely failed jobs (pending/unfinished excluded)"),
        )
        .arg(
            Arg::new("arch")
                .long("arch")
                .value_name("ARCH")
                .help("only jobs for this architecture"),
        )
        .arg(
            Arg::new("url_openqa")
                .long("url-openqa")
                .value_name("URL")
                .help("Override openQA URL (default: config openqa_instance)"),
        )
        .arg(
            Arg::new("url_dashboard_qam")
                .long("url-dashboard-qam")
                .value_name("URL")
                .help("Override QAM Dashboard base URL (default: derived from config)"),
        )
    }

    fn complete(&self, session: &Session, text: &str, _line: &str) -> Vec<String> {
        let mut out: Vec<String> = [
            "--all",
            "--failed",
            "--arch",
            "--url-openqa",
            "--url-dashboard-qam",
        ]
        .iter()
        .filter(|f| f.starts_with(text))
        .map(|s| (*s).to_owned())
        .collect();
        out.extend(template_completion(session, text));
        out
    }

    async fn call(&self, session: &mut Session, args: &ArgMatches) -> CommandResult {
        let rrid = require_update(session)?;

        let include_obsoleted = args.get_flag("all");
        let only_failed = args.get_flag("failed");
        let arch_filter = args.get_one::<String>("arch").cloned();

        let url_openqa = args
            .get_one::<String>("url_openqa")
            .cloned()
            .unwrap_or_else(|| session.config.openqa_instance.clone());
        let url_dashboard_qam = args
            .get_one::<String>("url_dashboard_qam")
            .cloned()
            .unwrap_or_else(|| {
                session
                    .config
                    .qem_dashboard_api
                    .trim_end_matches('/')
                    .strip_suffix("/api")
                    .unwrap_or(session.config.qem_dashboard_api.trim_end_matches('/'))
                    .to_owned()
            });

        let http = session
            .http_client()
            .map_err(|e| CommandError::Other(format!("could not build HTTP client: {e}")))?;

        // incident_id is maintenance_id (int for Maintenance, "1.2" for SLFO);
        // fall back to the review id in the SLFO case — mirrors openqa_overview.
        let effective_incident_id = if rrid.maintenance_id.parse::<i64>().is_ok() {
            rrid.maintenance_id.clone()
        } else {
            rrid.review_id.to_string()
        };

        let (build, _versions) =
            match oqa::get_incident_info(&http, &url_dashboard_qam, &effective_incident_id).await {
                Ok(v) => v,
                Err(e) => {
                    return Err(CommandError::Other(format!(
                        "QEM Dashboard query failed: {e}"
                    )));
                }
            };

        let mut jobs = match oqa::incident_jobs(&http, &build, &url_openqa, include_obsoleted).await
        {
            Ok(j) => j,
            Err(e) => {
                return Err(CommandError::Other(format!(
                    "openQA jobs query failed: {e}"
                )));
            }
        };
        if let Some(arch) = &arch_filter {
            jobs.retain(|j| &j.arch == arch);
        }
        // Count unfinished jobs before any --failed filtering so the "still
        // pending" message reflects reality regardless of the active filter.
        let pending_count = jobs.iter().filter(|j| is_pending(j)).count();
        if only_failed {
            jobs.retain(|j| {
                !is_pending(j)
                    && !PASSING.contains(&j.result.as_str())
                    && !NEUTRAL.contains(&j.result.as_str())
            });
        }

        if jobs.is_empty() {
            let msg = session
                .display
                .yellow(&format!("No openQA jobs for build {build:?}"));
            session.display.println(&msg);
            if pending_count > 0 {
                let pending = session
                    .display
                    .yellow(&format!("{pending_count} still pending"));
                session.display.println(&pending);
            }
            return Ok(());
        }

        let mut counts: BTreeMap<String, usize> = BTreeMap::new();
        for j in &jobs {
            *counts.entry(j.result.clone()).or_default() += 1;
        }
        let summary = counts
            .iter()
            .map(|(k, v)| format!("{k}={v}"))
            .collect::<Vec<_>>()
            .join(", ");
        session.display.println(&format!(
            "openQA jobs for build {build} ({}): {summary}",
            jobs.len()
        ));
        if pending_count > 0 {
            let pending = session
                .display
                .yellow(&format!("{pending_count} still pending"));
            session.display.println(&pending);
        }
        session.display.println("");
        for j in &jobs {
            let result = if is_pending(j) {
                session.display.yellow(&format!("{:<15}", j.result))
            } else if PASSING.contains(&j.result.as_str()) {
                session.display.green(&format!("{:<15}", j.result))
            } else if NEUTRAL.contains(&j.result.as_str()) {
                session.display.yellow(&format!("{:<15}", j.result))
            } else {
                session.display.red(&format!("{:<15}", j.result))
            };
            session
                .display
                .println(&format!("  {result} {:<8} {}  {}", j.arch, j.test, j.url));
        }
        Ok(())
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use crate::commands::testkit::{empty_session, matches, session_with_hosts};
    use wiremock::matchers::{method, path, query_param};
    use wiremock::{Mock, MockServer, ResponseTemplate};

    #[test]
    fn name_and_fanout_scope() {
        assert_eq!(OpenQAJobs.name(), "openqa_jobs");
        assert_eq!(OpenQAJobs.scope(), Scope::Fanout);
    }

    #[test]
    fn completion_offers_flags_and_templates() {
        let (session, _buf) = session_with_hosts("SUSE:Maintenance:1:1", &["h1"], "ok");
        let out = OpenQAJobs.complete(&session, "--f", "");
        assert_eq!(out, vec!["--failed"]);
    }

    #[tokio::test]
    async fn errors_when_no_report_loaded() {
        let (mut session, _buf) = empty_session();
        let args = matches(&OpenQAJobs, &[]);
        let err = OpenQAJobs.call(&mut session, &args).await.unwrap_err();
        assert!(matches!(err, CommandError::Other(_)));
    }

    /// Mounts a dashboard build lookup + an openQA jobs list on one server.
    async fn server_with_jobs(jobs: serde_json::Value) -> MockServer {
        let server = MockServer::start().await;
        Mock::given(method("GET"))
            .and(path("/api/incident_settings/1"))
            .respond_with(ResponseTemplate::new(200).set_body_json(serde_json::json!([
                {"settings": {"BUILD": "BUILD-42", "DISTRI": "sle", "VERSION": "15-SP5"}}
            ])))
            .mount(&server)
            .await;
        Mock::given(method("GET"))
            .and(path("/api/v1/jobs"))
            .and(query_param("build", "BUILD-42"))
            .respond_with(ResponseTemplate::new(200).set_body_json(jobs))
            .mount(&server)
            .await;
        server
    }

    #[tokio::test]
    async fn lists_jobs_with_summary() {
        let server = server_with_jobs(serde_json::json!({
            "jobs": [
                {"id": 1, "test": "install", "arch": "x86_64", "result": "passed", "state": "done", "clone_id": null},
                {"id": 2, "test": "boot", "arch": "x86_64", "result": "failed", "state": "done", "clone_id": null},
            ]
        }))
        .await;

        let (mut session, buf) = session_with_hosts("SUSE:Maintenance:1:1", &["h1"], "ok");
        let args = matches(
            &OpenQAJobs,
            &[
                "--url-dashboard-qam",
                &server.uri(),
                "--url-openqa",
                &server.uri(),
            ],
        );
        OpenQAJobs.call(&mut session, &args).await.unwrap();
        let out = buf.contents();
        assert!(out.contains("openQA jobs for build BUILD-42 (2)"), "{out}");
        assert!(out.contains("failed=1"), "{out}");
        assert!(out.contains("passed=1"), "{out}");
        assert!(out.contains("install"), "{out}");
        assert!(!out.contains("still pending"), "{out}");
    }

    #[tokio::test]
    async fn failed_filter_drops_passing_jobs() {
        let server = server_with_jobs(serde_json::json!({
            "jobs": [
                {"id": 1, "test": "install", "arch": "x86_64", "result": "passed", "state": "done", "clone_id": null},
                {"id": 2, "test": "boot", "arch": "aarch64", "result": "failed", "state": "done", "clone_id": null},
            ]
        }))
        .await;

        let (mut session, buf) = session_with_hosts("SUSE:Maintenance:1:1", &["h1"], "ok");
        let args = matches(
            &OpenQAJobs,
            &[
                "--failed",
                "--url-dashboard-qam",
                &server.uri(),
                "--url-openqa",
                &server.uri(),
            ],
        );
        OpenQAJobs.call(&mut session, &args).await.unwrap();
        let out = buf.contents();
        assert!(out.contains("failed=1"), "{out}");
        assert!(
            !out.contains("passed="),
            "passing job should be filtered: {out}"
        );
        assert!(!out.contains("install"), "{out}");
    }

    /// `--failed` must not list unfinished jobs (openQA `result:none`,
    /// non-terminal `state`) as failures, and must report them as pending.
    #[tokio::test]
    async fn failed_filter_excludes_pending_and_reports_count() {
        let server = server_with_jobs(serde_json::json!({
            "jobs": [
                {"id": 1, "test": "install", "arch": "x86_64", "result": "passed", "state": "done", "clone_id": null},
                {"id": 2, "test": "boot", "arch": "x86_64", "result": "failed", "state": "done", "clone_id": null},
                {"id": 3, "test": "kdump", "arch": "x86_64", "result": "none", "state": "scheduled", "clone_id": null},
            ]
        }))
        .await;

        let (mut session, buf) = session_with_hosts("SUSE:Maintenance:1:1", &["h1"], "ok");
        let args = matches(
            &OpenQAJobs,
            &[
                "--failed",
                "--url-dashboard-qam",
                &server.uri(),
                "--url-openqa",
                &server.uri(),
            ],
        );
        OpenQAJobs.call(&mut session, &args).await.unwrap();
        let out = buf.contents();
        // Only the genuinely failed job survives the filter.
        assert!(out.contains("boot"), "{out}");
        assert!(out.contains("failed=1"), "{out}");
        // The pending job is neither counted as failed nor listed.
        assert!(
            !out.contains("kdump"),
            "pending job must not be listed as failed: {out}"
        );
        // ... but it is reported as pending.
        assert!(out.contains("1 still pending"), "{out}");
    }

    /// A 5xx on the dashboard build lookup surfaces as Err, not an empty Ok the
    /// MCP surface would render as a bare success.
    #[tokio::test]
    async fn dashboard_failure_returns_err() {
        let server = MockServer::start().await;
        Mock::given(method("GET"))
            .respond_with(ResponseTemplate::new(500))
            .mount(&server)
            .await;

        let (mut session, _buf) = session_with_hosts("SUSE:Maintenance:1:1", &["h1"], "ok");
        let args = matches(
            &OpenQAJobs,
            &[
                "--url-dashboard-qam",
                &server.uri(),
                "--url-openqa",
                &server.uri(),
            ],
        );
        let err = OpenQAJobs.call(&mut session, &args).await.unwrap_err();
        assert!(matches!(err, CommandError::Other(_)));
    }

    /// A 5xx on the openQA jobs query (after a successful dashboard lookup) also
    /// surfaces as Err.
    #[tokio::test]
    async fn openqa_jobs_failure_returns_err() {
        let server = MockServer::start().await;
        Mock::given(method("GET"))
            .and(path("/api/incident_settings/1"))
            .respond_with(ResponseTemplate::new(200).set_body_json(serde_json::json!([
                {"settings": {"BUILD": "BUILD-42", "DISTRI": "sle", "VERSION": "15-SP5"}}
            ])))
            .mount(&server)
            .await;
        Mock::given(method("GET"))
            .and(path("/api/v1/jobs"))
            .respond_with(ResponseTemplate::new(500))
            .mount(&server)
            .await;

        let (mut session, _buf) = session_with_hosts("SUSE:Maintenance:1:1", &["h1"], "ok");
        let args = matches(
            &OpenQAJobs,
            &[
                "--url-dashboard-qam",
                &server.uri(),
                "--url-openqa",
                &server.uri(),
            ],
        );
        let err = OpenQAJobs.call(&mut session, &args).await.unwrap_err();
        assert!(matches!(err, CommandError::Other(_)));
    }

    /// In the default listing a pending job (`result:none`) is still shown and
    /// the pending count is surfaced. Color is stripped under `ColorMode::Never`,
    /// so the yellow-vs-red branch is exercised for coverage but not asserted on.
    #[tokio::test]
    async fn pending_job_listed_and_counted_in_default_listing() {
        let server = server_with_jobs(serde_json::json!({
            "jobs": [
                {"id": 1, "test": "kdump", "arch": "x86_64", "result": "none", "state": "running", "clone_id": null},
            ]
        }))
        .await;

        let (mut session, buf) = session_with_hosts("SUSE:Maintenance:1:1", &["h1"], "ok");
        let args = matches(
            &OpenQAJobs,
            &[
                "--url-dashboard-qam",
                &server.uri(),
                "--url-openqa",
                &server.uri(),
            ],
        );
        OpenQAJobs.call(&mut session, &args).await.unwrap();
        let out = buf.contents();
        assert!(out.contains("kdump"), "{out}");
        assert!(out.contains("1 still pending"), "{out}");
    }
}
